File: packages/playground/android/app/src/main/java/com/tiktok/sparkling/playground/PlaygroundNativeRouteRegistry.kt
// Copyright (c) 2025 TikTok Pte. Ltd.
// Licensed under the Apache License, Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.

package com.tiktok.sparkling.playground

import android.content.Context
import android.content.Intent

/**
 * Playground-only registry for `router.open({ scheme: "hybrid://native_route=<id>" })`.
 * Add new native pages by calling [register] (e.g. from Application) or by extending [installDefaultRoutes].
 */
object PlaygroundNativeRouteRegistry {

    const val NATIVE_ROUTE_SCHEME_PREFIX = "hybrid://native_route="

    private val handlers = mutableMapOf<String, (Context) -> Unit>()

    init {
        installDefaultRoutes()
    }

    fun installDefaultRoutes() {
        register("card_demo") { ctx ->
            ctx.startActivity(
                Intent(ctx, CardViewDemoActivity::class.java).apply {
                    addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                },
            )
        }
        register("debug_tool_switch") { ctx ->
            ctx.startActivity(
                Intent(ctx, DebugToolSwitchActivity::class.java).apply {
                    addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                },
            )
        }
    }

    fun register(routeId: String, handler: (Context) -> Unit) {
        handlers[routeId] = handler
    }

    fun unregister(routeId: String) {
        handlers.remove(routeId)
    }

    /**
     * @return true if [scheme] is a native-route URL and the route id was registered (even if [context] is null).
     */
    fun openIfMatches(scheme: String, context: Context?): Boolean {
        if (!scheme.startsWith(NATIVE_ROUTE_SCHEME_PREFIX)) {
            return false
        }
        val routeId = scheme.removePrefix(NATIVE_ROUTE_SCHEME_PREFIX).substringBefore("&")
        val handler = handlers[routeId] ?: return false
        context?.let { handler(it) }
        return true
    }
}

File: packages/playground/ios/SparklingGo/SparklingGo/MethodServices/PlaygroundNativeRouteRegistry.swift
// Copyright 2025 The Sparkling Authors. All rights reserved.
// Licensed under the Apache License, Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.

import Foundation
import UIKit
import SparklingMethod

/// Playground-only registry for `router.open({ scheme: "hybrid://native_route=<id>" })`.
/// Add new native pages with `register(_:handler:)`.
final class PlaygroundNativeRouteRegistry {

    static let shared = PlaygroundNativeRouteRegistry()

    static let nativeRouteSchemePrefix = "hybrid://native_route="

    typealias NativeRouteHandler = (@escaping SparklingMethod.PipeMethod.CompletionBlock) -> Void

    private var routes: [String: NativeRouteHandler] = [:]

    private init() {
        installDefaultRoutes()
    }

    func installDefaultRoutes() {
        register("card_demo") { [weak self] completion in
            self?.presentRoot(CardViewDemoViewController(), completion: completion)
        }
        register("debug_tool_switch") { [weak self] completion in
            self?.presentRoot(DebugToolSwitchViewController(), completion: completion)
        }
    }

    func register(_ routeId: String, handler: @escaping NativeRouteHandler) {
        routes[routeId] = handler
    }

    func unregister(_ routeId: String) {
        routes.removeValue(forKey: routeId)
    }

    /// - Returns: `true` if the scheme is a native-route URL and `routeId` was registered.
    func tryOpen(scheme: String, completion: @escaping SparklingMethod.PipeMethod.CompletionBlock) -> Bool {
        guard scheme.hasPrefix(Self.nativeRouteSchemePrefix) else {
            return false
        }
        let rest = String(scheme.dropFirst(Self.nativeRouteSchemePrefix.count))
        let routeId = rest.split(separator: "&", maxSplits: 1, omittingEmptySubsequences: false).first.map(String.init) ?? ""
        guard let handler = routes[routeId] else {
            return false
        }
        handler(completion)
        return true
    }

    private func presentRoot(_ viewController: UIViewController, completion: @escaping SparklingMethod.PipeMethod.CompletionBlock) {
        guard let topVC = Self.topViewController() else {
            completion(.failed(message: "No view controller to present from"), nil)
            return
        }
        if let nav = topVC.navigationController {
            nav.pushViewController(viewController, animated: true)
        } else {
            let nav = UINavigationController(rootViewController: viewController)
            nav.modalPresentationStyle = .fullScreen
            topVC.present(nav, animated: true)
        }
        completion(.succeeded(), nil)
    }

    private static func topViewController(base: UIViewController? = nil) -> UIViewController? {
        let root = base ?? UIApplication.shared.connectedScenes
            .compactMap { ($0 as? UIWindowScene)?.keyWindow?.rootViewController }
            .first
        if let nav = root as? UINavigationController {
            return topViewController(base: nav.visibleViewController)
        }
        if let presented = root?.presentedViewController {
            return topViewController(base: presented)
        }
        return root
    }
}
